main/src/ca/uwaterloo/flix/language/phase/ConstraintGeneration.scala

```diff
@@ -0,0 +1,1071 @@
+                val (actualTpe, actualEff) = visitExp(body)
+
+                // unify the operation return type with its tvar
+                c.unifyTypeM(actualTpe, opTvar, body.loc)
+
+                // unify the handler result type with the whole block's tvar
+                c.unifyTypeM(tvar, actualTpe, body.loc)
+                (actualTpe, actualEff)
+            }
+        }
+
+        // TODO ASSOC-TYPES The types used here are not correct.
+        // TODO ASSOC-TYPES We should use set subtraction instead.
+        val (tpe, bodyEff) = visitExp(exp)
+        val correctedBodyEff = c.purifyEff(effUse.sym, bodyEff) // Note: Does not work for polymorphic effects.
+        val (_, effs) = rules.map(visitHandlerRule(_, tpe)).unzip
+        c.unifyTypeM(continuationEffect, Type.mkUnion(correctedBodyEff :: effs, loc), loc)
+        c.unifyTypeM(tvar, tpe, loc)
+        val resultTpe = tpe
+        //        val resultEff = continuationEffect
+        val resultEff = Type.freshVar(Kind.Eff, loc) // TODO ASSOC-TYPES should be continuationEffect
+        (resultTpe, resultEff)
+
+      case Expr.Do(op, args, tvar, loc) =>
+        val effect = root.effects(op.sym.eff)
+        val operation = effect.ops.find(_.sym == op.sym)
+          .getOrElse(throw InternalCompilerException(s"Unexpected missing operation $op in effect ${op.sym.eff}", loc))
+        val effTpe = Type.Cst(TypeConstructor.Effect(op.sym.eff), loc)
+
+        def visitArg(arg: KindedAst.Expr, fparam: KindedAst.FormalParam): Type = {
+          val (tpe, eff) = visitExp(arg)
+          c.expectTypeM(expected = fparam.tpe, actual = tpe, arg.loc)
+          eff
+        }
+
+        // We special case the result type of the operation.
+        val operationType = operation.spec.tpe.typeConstructor match {
+          case Some(TypeConstructor.Void) =>
+            // The operation type is `Void`. Flix does not have subtyping, but here we want something close to it.
+            // Hence we treat `Void` as a fresh type variable.
+            // An alternative would be to allow empty pattern matches, but that is cumbersome.
+            Type.freshVar(Kind.Star, operation.spec.tpe.loc, false, Ast.VarText.Absent)
+          case _ => operation.spec.tpe
+        }
+
+        // length check done in Resolver
+        val effs = (args zip operation.spec.fparams) map {
+          case (arg, fparam) => visitArg(arg, fparam)
+        }
+
+        c.unifyTypeM(operationType, tvar, loc)
+        val resTpe = tvar
+        val resEff = Type.mkUnion(effTpe :: operation.spec.eff :: effs, loc)
+
+        (resTpe, resEff)
+
+      case Expr.InvokeConstructor(constructor, args, loc) =>
+        val classTpe = getFlixType(constructor.getDeclaringClass)
+        val (_, _) = args.map(visitExp).unzip
+        val resTpe = classTpe
+        val resEff = Type.IO
+        (resTpe, resEff)
```

Comment:
BIG TEXT: I would think that every case would include calls to c.unify unless there is a good reason (e.g. constant). 

Could we double check that? If there is no such call it should be documented imho! "This expression does not generate constraints because ...".

Presumably getFlixType does not generate constraints!

Do we not need to match the arguments against the types of the constructor? Am I stupid?